### PR TITLE
mavlink: use RADIO_STATUS to regulate parameter sending

### PR DESCRIPTION
**Pull request by brad112358**
This avoids parameter message loss because of radio buffer overrun.

Fix a long standing issue with parameter download overrunning the radio's transmit FIFO on slow radio links with no hardware flow control, but, which utilize RADIO_STATUS messages to provide txbuf usage info.

Make the _radio_status_critical bool available via a new predicate method and use it where we check the serial port's transmit buffer,_mavlink->get_free_tx_buf() in the parameter download code.  Note, that because we allow parameter messages to use available space up to RADIO_BUFFER_CRITICAL_LOW_PERCENTAGE, other message streams will have their rate gradually reduced during parameter download by update_radio_status().  This is intentional and desirable.  I also include a change to this function to ensure that the rate multiplier doesn't get too far from it's normal range.  These bounds help speed up recovery to the best rate once parameter download completes.

This strategy is less disruptive and simpler than rewriting buik parameter download code to use the rate multiplier based stream flow control used for most MAVLink messages.

This change was tested with QGroundControl over the TBS Crossfire radio system with several released and unreleased versions of the Crossfire firmware.  It improved parameter download time from about 40 seconds to 20-25 seconds by eliminating dropped messages in the first phase of parameter download so few or no additional individual parameter requests were needed.  Without this fix, most messages in the first phase of the bulk parameter download are dropped by the Crossfire receiver and needed to be fetched by QGroundControl individually, slowing down parameter download and reducing it's reliability.

See [issue 10618](https://github.com/PX4/PX4-Autopilot/issues/10618) and [QGC issue 6890](https://github.com/mavlink/qgroundcontrol/issues/6890) for some past discussion.

### CODE REVIEW

File: src/modules/mavlink/mavlink_main.cpp [hunk 1]
```diff
@@ -1476,6 +1476,11 @@ Mavlink::update_radio_status(const radio_status_s &radio_status)
 			/* this indicates spare bandwidth, increase by 2.5% */
 			_radio_status_mult *= 1.025f;
 		}
+
+		/* Constrain radio status multiplier between 1% and 100% to allow recovery */
+		if (_radio_status_mult > 1.0f) { _radio_status_mult = 1.0f; }
```
Comment (dagar) [inline src/modules/mavlink/mavlink_main.cpp:1481]: _radio_status_mult
```suggestion
		_radio_status_mult = math::constrain(_radio_status_mult, 0.01f, 1.0f);
```

**Comment by dagar** [comment]
Thanks for finally doing something about this! We should also do a quick verification on the SiK radios that didn't have flow control.

File: src/modules/mavlink/mavlink_main.h [hunk 2]
```diff
@@ -518,6 +518,8 @@ class Mavlink final : public ModuleParams
 
 	static hrt_abstime &get_first_start_time() { return _first_start_time; }
 
+	bool radio_status_critical() { return _radio_status_critical; }
```
Comment (dagar) [inline src/modules/mavlink/mavlink_main.h:521]:
```suggestion
	bool radio_status_critical() const { return _radio_status_critical; }
```

File: src/modules/mavlink/mavlink_main.cpp (hunk 1 above)
Comment (brad112358) [reply src/modules/mavlink/mavlink_main.cpp:1481]: These suggestions seem to be a new feature.  I'd just click commit suggestion, but it looks like you missed deleting the first conditional

File: src/modules/mavlink/mavlink_main.h (hunk 2 above)
Comment (brad112358) [reply src/modules/mavlink/mavlink_main.h:521]: OK.  This one is good.  I'll try the commit suggestion feature here...

File: src/modules/mavlink/mavlink_main.cpp [hunk 3]
```diff
@@ -1476,6 +1476,11 @@ Mavlink::update_radio_status(const radio_status_s &radio_status)
...
+
+		if (_radio_status_mult < 0.01f) { _radio_status_mult = 0.01f; }
```
Comment (dagar) [inline src/modules/mavlink/mavlink_main.cpp:1483]:
```suggestion
```

File: src/modules/mavlink/mavlink_main.cpp [hunk 4]
```diff
@@ -1476,6 +1476,11 @@ Mavlink::update_radio_status(const radio_status_s &radio_status)
 			/* this indicates spare bandwidth, increase by 2.5% */
 			_radio_status_mult *= 1.025f;
 		}
+
+		/* Constrain radio status multiplier between 1% and 100% to allow recovery */
+		if (_radio_status_mult > 1.0f) { _radio_status_mult = 1.0f; }
+
```
Comment (dagar) [inline src/modules/mavlink/mavlink_main.cpp:1482]:
```suggestion
```

**Review by dagar — APPROVED** [review]

**Comment by brad112358** [comment]
Thanks for merging this so quickly!  Any thoughts on how long it might take to make it's way to a stable release?  I've been working with TBS in an effort to encourage them to improve their MAVLink support over Crossfire.  They had implemented several undocumented and horrible kludges, probably trying to make things work with these FC flow control bugs (I also fixed parameter download flow control in Ardupilot).  Their receiver transmit FIFO is way too big compared to the data rate, they rate limit packets based on specific message type and parameter download state, removing the ability for the GS and FC to set effective policy on message rates.  The combination of message discard for most message types and large FIFO means that RADIO_STATUS never reports less than 99% free, no matter how fast the FC sends messages, except during parameter download (which also didn't work correctly because of the issue fixed here).  I hope that once PX4 and Ardupilot both handle RADIO_STATUS correctly, I might convince them to pull all of the kludges out and then latency and reliability should be much better on Crossfire.  Perhaps I can even convince them to allow MAVLink messages to flow in their longest range 50Hz radio mode.  With proper flow control, MAVLink should be usable even when switching between 50Hz and 150Hz radio modes.

**Comment by dagar** [comment]
That's great to hear @brad112358. I wanted Crossfire to be my default solution for development (rock solid RC + telemetry), but I never got good mavlink performance (not to mention bluetooth annoyances).

> Any thoughts on how long it might take to make it's way to a stable release?

We're trying to close in on a v1.12 release imminently (weeks), but there are still a number of outstanding items to resolve. Any testing you can do with current PX4 master would get us there sooner.

Please let me know if there's anything else we can do on the PX4 side.

**Comment by brad112358** [comment]
@dagar, Thanks for the info! Yes, at just 1 gram and 2 wires for RC + 1200B/s MAVLink, Crossfire really has huge potential with PX4 or Ardupilot.  It should be ideal for my use case of 190g all up weight autonomous FPV camera builds which fly for 32 minutes with 4K video recording.  I've been pushing TBS hard and they have made a lot of progress fixing the issues I've reported.  Bluetooth disconnect issues have been fixed in their ALPHA and BETA releases, as have WiFi connectivity and several minor MAVLink issues.  The latest test releases their developer sent me were mostly working correctly without the old kludges, but he seems to have stopped work on it for some reason and hasn't responded to my feedback for a couple of weeks.  If it was open source, I'd have fixed all the issues myself by now with all the time I've spent testing for them.  If you know anyone with connections at TBS, it couldn't hurt to remind them that they could expand their market a little if they finally make MAVLink over crossfire work as well as it should.  Thanks!

**Pull request merged**